### DISCUSSION TITLE
Optional URL parameters

### DISCUSSION THREAD

**Discussion by chaance** [remix-run/react-router#9208] [Proposals]
In v5, a dynamic URL parameter could be optional if a trailing `?` is added to the `Route` component's `path`:

```jsx
<Route path="/blog/:postId/:action?" component={BlogPost} />
```

This doesn't work quite the same in v6. To achieve something similar, you'd need to use nested routes with `<Outlet />`:

```jsx
<Route path="/blog" element={<BlogPostLayout />}>
  <Route path=":action" element={<BlogPost />} />
  <Route path="" element={<BlogPost />} />
</Route>
```

```jsx
function BlogPostLayout() {
  let context = useSomeSharedLogic();
  return <Outlet context={context} />;
}
```

Many users have expressed that this change requires some significant refactoring to their apps, and that bringing back support for the v5 API for optional params would be a simpler solution for them. [See the original issue here.](https://github.com/remix-run/react-router/issues/7285)

To consider this potential change, I'd start by asking: would simply changing the way the v6 matching algorithm works with a trailing `?` potentially break existing v6 apps?

> **Note:** Please keep the comments kind and productive. This is not a space to vent about changes you don't like, but to help create solutions for yourself and others. Off-topic or overly negative comments may be removed at any time.

**Comment by timdorr** [comment]
This is now superseded by #9546

**Comment by ryanflorence** [comment]
Locked in favor of ^
